### ISSUE TITLE
`@typescript-eslint/utils` missing types for report `message` and invalid test case `message`

### ISSUE THREAD

**Issue by Pearce-Ropion** [typescript-eslint/typescript-eslint#4917]
This issue is for `@typescript-eslint/utils` but there wasn't an option for that package in the new issue screen.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

`@typescript-eslint/utils` is missing support for the [`message` property](https://eslint.org/docs/developer-guide/working-with-rules#contextreport). The relevant types that should be updated would be:
* `ReportDescriptor` in `TSESLint.RuleContext['report']`
* `TestCaseError` in `TSESLint.InvalidTestCase`.

In both cases, either `messageId` OR `message` should be required (currently only `messageId` is required).

**Repro**
[Typescript playground repro](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgUQMoBlgDsYFUbAA2AzgDRwCCqAKgPoByA8gCLK3UCaACmnAL5wAZlAgg4AcgACMAJ5gApsQDGUYGBgBaRYWwwA9AFcCJcQG4AUOcEGsSghCxwA5vJgBhUWELyAHgFlFYgBDFwAKAEoALjhiGFUsJ0RzODgoVwMoRwAiJU9vHzgQQJD5LIs+S115KEEgpXk4RnVgB2JEPgtZBTgA4mCXAEkAEzaAXgkivpKNIPE4AB8J4pcNACMzS1ysWLgVeSCYeQAlA284cbRMHHwiYgA6E+83NIPoUKwgovOAPjgAAwAFjAYGBiJE9HpfJ8vPI7rkQHooKd5HoACQID5FPh-cIWcy+SCwXateBIs7jPYHY7IgA8AG0mvZtgBdci9fryYbEb6hBDJOCY+TRcTEUTyDRk8WC8SkflFGBBaJ8lIpLpCiTEAxOFyxFpYGX8lJDCBKMFJFUqoaKFRqJnC0VfK3KVTNBwGi0pNLwopYK1DaK1EjyWUWvghlXKAHyECKuB01mGwrLRRKxMpcSTDkzcT20AwpNTFyUd0ejPJtY5jV5s6ZkpwABCJZSFWb4cph2iW0OPngo1+vMTXBETign2iWAgVp+5o9Wx2DvkHl9wCZ1ABB1QAIgpyG9c5WAAbkEdENznA4gZ5HiPXBgII4KEF0uhiu9WuN1ud3uBofj8AhuEM43l2vgwHcaSEjAA43haE5WuGMEFhy0QuO4eS+OyJQRAhHp8OEab8ImLb8OE4ZWrUpwwIyepmnSCB8AmeF4nOpLItQiiHFAZ5YPIADuKAYLoNwkA8bEcdU0FwGAQRQMQ1TCtIcjWi6mjaLoejSbJ1QGkxliSuxsTVOBNihCKYoSsiGjSuQkrkMqcBHie0TxuG2COf+zmJvZsGfOqZlfG5f6nocsRNiquRWsKLFwEEZ4AIymGFKTVCIsmeYh3kwWq0RUHQTCsOw3BoHcQ4QCOnw4TetYuChrgeOA+SYWEpEEc2BHMkR-IdXhQA)

```TS
import { ESLintUtils, AST_NODE_TYPES } from '@typescript-eslint/utils';

function getComplexMessage(): string {
  return "complex message";
}

interface Options {};
type MessageIds = 'message-a' | 'message-b';

const createRule = ESLintUtils.RuleCreator(name => `https://example.com/rule/${name}`);

export const rule = createRule<[Options], MessageIds>({
  name: 'some-rule-name',
  meta: {
    type: 'suggestion',
    docs: {
      description: 'some description',
      recommended: false,
    },
    schema: [],
    messages: {
      'message-a': 'simple message A',
      'message-b': 'simple message B',
    }
  },
  create: context => ({
    Program: node => {
      const someConditionThatShouldBeInvalid = true;

      if (someConditionThatShouldBeInvalid) {
        context.report({
          node,
          // Object literal may only specify known properties, but 'message' does not exist in type 'TestCaseError<MessageIds>'. Did you mean to write 'messageId'?
          message: getComplexMessage(),
        })
      }
    }
  }),
  defaultOptions: [{}],
});

const ruleTester = new ESLintUtils.RuleTester({
  parser: '@typescript-eslint/parser',
});

ruleTester.run('some-rule-name', rule, {
  valid: [],
  invalid: [
    {
      name: 'some invalid test',
      code: 'const a = 1;',
      errors: [
        {
          type: AST_NODE_TYPES.Program,
          // Object literal may only specify known properties, but 'message' does not exist in type 'TestCaseError<MessageIds>'. Did you mean to write 'messageId'?
          message: getComplexMessage(),
        }
      ]
    }
  ]
})
```

**Expected Result**

Using the `message` property in the report function or in an invalid test case error should not throw a typescript error. `messageId` should not be required if using a `message` property.

**Actual Result**

The `report` function and invalid test case errors should allow using the `message` property. If `message` is defined, do not require a `messageId`.

**Additional Info**

Both eslint and the test runner work correctly when using the `message` property and the errors are silenced with `@ts-ignore`. It is just the type definitions that are missing the `message` property.

**Versions**

| package                         | version |
| ------------------------------- | ------- |
| `@typescript-eslint/utils` | `5.22.0` |
| `TypeScript`                    | `4.6.3` |
| `node`                          | `16.13.1` |

**Comment by bradzacher** [comment]
This is done on purpose because using raw message strings is generally considered bad practice.

Raw message strings are ugly to test because you have to copy paste the entire string to get it to exactly match. There are no types we can create which let you validate the tests are correct at type check time if you are relying on messages.

OTOH messageIds are typesafe and easily testable - the opposite of raw message strings.

Our recommendation is to use messageIds for everything. You can leverage [placeholders](https://eslint.org/docs/developer-guide/working-with-rules#using-message-placeholders) to construct complex messages - which is much preferable to using manually constructed strings.

This is the way the types have been since their creation some 4 years ago, and this is how we will continue to maintain them.
